p2p/src/authenticated/actors/tracker/actor.rs

```diff
@@ -386,42 +283,35 @@ impl<E: Spawner + Rng + Clock + GClock + Metrics, C: Scheme> Actor<E, C> {
         true
     }
 
-    fn handle_peers(&mut self, peers: wire::Peers) -> Result<(), Error> {
+    fn handle_peers(&mut self, peers: Vec<types::SignedPeerInfo<C>>) -> Result<(), Error> {
         // Ensure there aren't too many peers sent
-        let peers_len = peers.peers.len();
-        if peers_len > self.peer_gossip_max_count {
-            return Err(Error::TooManyPeers(peers_len));
+        if peers.len() > self.peer_gossip_max_count {
+            return Err(Error::TooManyPeers(peers.len()));
         }
 
         // We allow peers to be sent in any order when responding to a bit vector (allows
         // for selecting a random subset of peers when there are too many) and allow
         // for duplicates (no need to create an additional set to check this)
         let mut updated = false;
-        for peer in peers.peers {
-            // Check if address is well formatted
-            let address = socket_from_payload(&peer)?;
-
+        for peer in peers {
             // Check if IP is allowed
-            let ip = address.ip();
+            let ip = peer.socket.ip();
             if !ip::is_global(ip) && !self.allow_private_ips {
                 return Err(Error::PrivateIPsNotAllowed(ip));
             }
 
-            // Check if peer is signed
-            let signature = peer.signature.as_ref().ok_or(Error::PeerUnsigned)?;
-
-            // Check if public key is well-formatted and if peer is us
-            let public_key = C::PublicKey::try_from(&signature.public_key)
-                .map_err(|_| Error::InvalidPublicKey)?;
-            if public_key == self.crypto.public_key() {
+            // Check if peer is us
+            if peer.public_key == self.crypto.public_key() {
                 return Err(Error::ReceivedSelf);
             }
 
             // If any signature is invalid, disconnect from the peer
-            let signature = C::Signature::try_from(&signature.signature)
-                .map_err(|_| Error::InvalidSignature)?;
-            let payload = wire_peer_payload(&peer);
-            if !C::verify(Some(&self.ip_namespace), &payload, &public_key, &signature) {
+            if !C::verify(
```

Comment:
If we already have `<C: Scheme>` associated with `SignedPeerInfo`, I think we should implement a `verify` method on it (and keep this logic inside the type).